modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/transaction/XAExtensionTransactionalResource.java

```diff
@@ -0,0 +1,168 @@
+
+import static javax.transaction.xa.XAResource.TMSUCCESS;
+import org.mule.runtime.api.connection.ConnectionHandler;
+import org.mule.runtime.core.api.transaction.Transaction;
+import org.mule.runtime.core.api.transaction.TransactionException;
+import org.mule.runtime.core.config.i18n.CoreMessages;
+import org.mule.runtime.core.transaction.IllegalTransactionStateException;
+import org.mule.runtime.core.transaction.TransactionCoordination;
+import org.mule.runtime.core.transaction.XaTransaction;
+import org.mule.runtime.extension.api.connectivity.XATransactionalConnection;
+
+import javax.transaction.xa.XAResource;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Specialization of {@link ExtensionTransactionalResource} capable of joining XA transactions.
+ *
+ * @param <T> the generic type of the {@link XATransactionalConnection} that will join the transaction
+ * @since 4.0
+ */
+public class XAExtensionTransactionalResource<T extends XATransactionalConnection> extends ExtensionTransactionalResource<T> implements XaTransaction.MuleXaObject
+{
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(XAExtensionTransactionalResource.class);
+
+    private XAResource enlistedXAResource;
+
+    private volatile boolean reuseObject = false;
+
+    /**
+     * {@inheritDoc}
+     */
+    public XAExtensionTransactionalResource(T connection, ConnectionHandler<T> connectionHandler, Transaction transaction)
+    {
+        super(connection, connectionHandler, transaction);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean enlist() throws TransactionException
+    {
+        XaTransaction transaction = getTransaction();
+
+        synchronized (this)
+        {
+            if (!isEnlisted())
+            {
+                final XAResource xaResource = getConnection().getXAResource();
+                boolean wasAbleToEnlist = transaction.enlistResource(xaResource);
+                if (wasAbleToEnlist)
+                {
+                    enlistedXAResource = xaResource;
+                }
+            }
+        }
+
+        return isEnlisted();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean delist() throws Exception
+    {
+        if (!isEnlisted())
+        {
+            return false;
+        }
+
+        XaTransaction transaction = getTransaction();
+
+        synchronized (this)
+        {
+            if (isEnlisted())
+            {
+                boolean wasAbleToDelist = transaction.delistResource(enlistedXAResource, TMSUCCESS);
+                if (wasAbleToDelist)
+                {
+                    enlistedXAResource = null;
+                }
+            }
+            return !isEnlisted();
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void close() throws Exception
+    {
+        try
+        {
+            getConnection().close();
+        }
+        catch (Exception e)
```

Comment:
this and other things like the synchronized blocks and stuff are patterns already present in existing XA enabled classes such as ConnectorWrapper or BitronixConnectorWrapper. Because all of them have the same pattern, I take it these are somehow correct and necessary. 

They can however be reviewed when we revisit transactions as a whole at the runtime level